### ISSUE TITLE
enable browsing in cloud storage

### ISSUE THREAD

**Issue by hannahfrost** [projecthydra-labs/hyku#903]
Is Hyku missing Browse Everything?

**Comment by jcoyne** [comment]
No, but it's only configured for "File System"

<img width="283" alt="screen shot 2017-04-04 at 1 06 27 pm" src="https://cloud.githubusercontent.com/assets/92044/24671580/97ac8638-1937-11e7-8f4b-5521861c976e.png">

**Comment by jcoyne** [comment]
You configure the providers with this file, so we'd need to write some sort of adapter that could hold the keys for each tenant: https://github.com/projecthydra-labs/hyku/blob/master/config/browse_everything_providers.yml

**Comment by atz** [comment]
Looks like we could crudely but easily monkeypatch this method:
https://github.com/projecthydra/browse-everything/blob/d0438f114a9b5c0a12fa6a5b493cb41d31255019/lib/browse_everything.rb#L38-L43

We could also do the proper OO thing and extend browse-everything and override `config`.  

But in any case, `configure` can already take a hash, in which case it doesn't try to read the yaml. This is very doable.

**Comment by hannahfrost** [comment]
Sounds like we have a plan here. Ready for the Ready Column, @atz @jcoyne @mjgiarlo?

**Comment by mjgiarlo** [comment]
@hannahfrost Do we want to take this on prior to the beta launch on 5/15, or prioritize it to happen during the beta after we've focused on QA/fixes?

**Comment by hannahfrost** [comment]
@mjgiarlo I'm good with making it a priority during the beta and after the current round of QA/fixes

**Comment by mjgiarlo** [comment]
@hannahfrost OK! Me too.

**Comment by hannahfrost** [comment]
Also do we need a place in the UI to configure this? Design needed?

**Comment by mjgiarlo** [comment]
@hannahfrost Ah, yes, good call.

**Comment by ggeisler** [comment]
@jcoyne @mjgiarlo A couple of questions before I can do the design:

* This is a Hyku ticket but should this be done in Hyrax? Just asking because I see a `browse_everything_providers.yml` in Hyrax.

* Can you configure and activate more than one provider at a time?

**Comment by mjgiarlo** [comment]
@ggeisler

Yes, this should be done in Hyrax, and you can configure and activate more than one provider at a time, IIRC.

**Comment by ggeisler** [comment]
Here is a first-pass at a mockup of how we might provide UI-based configuration of BrowseEverything. I've no experience with BrowseEverything so it's possible I've misunderstood things; the design here is basically just translating the `browse_everything_providers.yml` file to a UI. 

I suppose it is possible that an easier first cut at this might be just to add BrowseEverything to the Settings table (first mockup below only) and just provide the on/off toggle to control whether the feature is presented to the end-user, and if it is set to On, then we still read the `browse_everything_providers.yml` file for provider configuration? In any case, the mockups below represent the goal of both enabling the feature and configuring providers through the UI.

### Add BrowseEverything to the Administration > Settings page
Adds the BrowseEverything setting to the bottom of the table. The BrowseEverything entry is similar to other features in the table, except if we want to enable provider configuration through the UI, then we add a "Configure" toggle to give access the provider settings.

![menu-setting](https://cloud.githubusercontent.com/assets/101482/25974093/f571fe4c-365b-11e7-911b-ef8f0c1670de.png)

### Provider configuration
Clicking the "Configure" toggle expands the page to reveal the provider configuration. This is a pretty standard form. The important things here are validation. Ideally, the user should not be able to save the form unless each checked provider is valid. If it isn't feasible to validate the entries for a provider, we should at least validate that each required field for a checked provider is not null before saving the form. (Ignore the horizontal line below the S3 optional heading; just a mockup artifact.)

![configuration](https://cloud.githubusercontent.com/assets/101482/25974205/a2c7a2fe-365c-11e7-8b4c-e96cca57d41b.png)

**Comment by ggeisler** [comment]
Below is an alternate way to present the page, if adding the configuration panel to the existing Settings page seems unwieldy (it is the only setting that has this expand/collapse section, which might be a little weird).

This assumes we reconfigure the current Settings page to allow secondary settings-related pages, similar to Hyku (see sidebar of mockup). (I think that would be a good idea, regardless of how we decide to implement this Browse Everything/Storage Providers page.)

Basically, give Storage Providers a dedicated settings page, with a tab for each provider. In each provider's tab section, present the same options as in the mockup above.

![display-set-mockups_graffle__admin__cloud_storage__browse_everything_](https://cloud.githubusercontent.com/assets/101482/26516619/3d916b52-423e-11e7-8f41-8b0afde4a259.png)

**Comment by dbernstein** [comment]
storage provider tabs ++

**Comment by mjgiarlo** [comment]
We decided to work on this immediately after returning from OR2017.

**Comment by mjgiarlo** [comment]
Needed: how to encrypt and handle users' credentials.

**Comment by hannahfrost** [comment]
Lots of good stuff here in this thread. Needs to be re-expressed in a new ticket and added to the Hyrax backlog @vantuyls

**Comment by vantuyls** [comment]
https://github.com/samvera/hyrax/issues/2110
